Installation && Exectution Optimization.

clearloop opened the pull request:
# Features

`./cita` directly.

# Optimization

### 1. Node 'test-chain/0' not responding to pings
> if not `daemon.sh bin/cita start test-chain/0`, this bug occurs.

Added a condition into `cita/env.sh`, if run `start` command, daemon automatically

### 2. Unable to resolve host linuxkit-025000000001
> mac environment should set `--net=bridge`

Added a condition into `cita/env.sh`, if on macOS, the tag will move to bridge.

### 3. Node Directory Not Found
> Should run `bin/cita \<command\>' while operating node.

Replaced `pwd` as `$0`, to fix the scripts' route.

[1]: https://docs.citahub.com/en-US/cita/getting-started
[2]: https://github.com/cryptape/cita/releases/tag/v0.21
[gist]: https://gist.github.com/clearloop/c8a292c181aacbca33e63f445d85b745#file-cita-sh

kaikai1024 commented:
Please rebase the latest develop branch.

clearloop commented:
> Please rebase the latest develop branch.

Okay, it took me a while XD

clearloop commented:
Should I back to 2018....

Crazy Night🤦🏻‍♂️

kaikai1024 commented:
@leeyr338

ouwenkg commented on Makefile:
```diff
@@ -2,6 +2,9 @@ CARGO=RUSTFLAGS='-F warnings' cargo
 
+citaship:
+	make -f ./scripts/Makefile
+
```
In your branch situation, other commands related with `cargo` in this makefile , such as `test`, `bench`  also need to be changed.

clearloop replied:
Okay, this commit terms to solve the starters' self compilation.

Compare with the code developers who need `test`, `bench` directly, once they test the code of the whole project, use `cargo` might be more in line with their habits than `make`?_

__I'm not sure about it now__, because, in the same situation, not every developer needs to compile the code by him/herself,  so I wrote theme like above...

+ Completely separated docker operation and local operation.
+ If don't want to operate the docker container, `make test`, `make bench`, are still work.
+ If developer want to `make test`, `make bench` in the docker container, `./scripts/env.sh make test`, `./scripts/env.sh make bench` will help.
+ Compile release code, just `make` in the source dir, this might be the most commonly used.

leeyr338 replied:
I found it is building very slowly in a docker container. It takes me about 1h to build it from the beginning.

clearloop replied:
> I found it is building very slowly in a docker container. It takes me about 1h to build it from the beginning.

Yep, 40+mins once, my mac blowed up last night when I built docker container、Docerfile、release version together...

The best why to install might be using the __compiled `code`__ or a __complete `docker image`__?

leeyr338 replied:
For a newcomer, he/she may prefer to `run` CITA first, rather than compiling it from source, or even learning the code.
So, If user `install` CITA, I think is better to use the compiled code or a complete docker image (I think a complete docker image is better). And,
If user `clone` CITA, that means he/she is ready to hack the code, so we need to help him/her `compile & run` from `code`.
What do you think about it?

leeyr338 replied:
You mean `rm docker container` While building in scripts can get a better performance?

kaikai1024 replied:
yeah, the first choice should be friend for uses

clearloop replied:
yep, I'm agree with your opinion, so:

|Mode|Description|needs|
|-|-|-|
|Run|Compiled code, with a `cita` command can run in `$PATH`|brew/snap|
|Install|Just pull a `docker image`, a `env.sh` is all they need.|Complete Docker Image|
|Clone|`env.sh make release` doesn't a complex way, they're ready for it.|keep|

1. plz check my formula~ I have just optimized it last night, it's quite nice~ 
2. If the formula is ok, I'm trying to write a Docker Image now.
3. For `make`, I'll keep it as it is ~

kaikai1024 replied:
nice job!

leeyr338 replied:
I got some errors here:
```\$ brew tap clearloop/hyperfox
Updating Homebrew...
==> Tapping clearloop/hyperfox
Cloning into '/usr/local/Homebrew/Library/Taps/clearloop/homebrew-hyperfox'...
remote: Repository not found.
fatal: repository 'https://github.com/clearloop/homebrew-hyperfox/' not found
Error: Failure while executing; `git clone https://github.com/clearloop/homebrew-hyperfox /usr/local/Homebrew/Library/Taps/clearloop/homebrew-hyperfox --depth=1` exited with 128.
```

clearloop replied:
oops, I wrote the wrong url, 

__try this: `tap clearloop/homebrew-hypercask`__

clearloop replied:
Wait a minute, something occurs...I'll fix it soon XD

leeyr338 replied:
Thanks, it works!
If we have another two features in `cita` script, that would be more perfect:
- Generating chain configuration, just likes:
```
./env.sh ./scripts/create_cita_config.py create --super_admin "0x4b5ae4567ad5d9fb92bc9afd6a657e6fa13a2523"  --chain_name test2-chain --jsonrpc_port 2337 --ws_port 5337 --grpc_port 6000 --nodes "127.0.0.1:8000,127.0.0.1:8001,127.0.0.1:8002,127.0.0.1:8003"
```
- Exposing a port from the running docker which likes `-p 1337:1337`.

After that, we can get a `cita-block-chain` by using `cita` which just installed.

leeyr338 replied:
And then, the user can use a client (sdk, cita-cli, etc.) to access the `block-chain`, do anything they want (check the status of the chain, deploy a contract on to or something else).

clearloop replied:
> If we have another two features in `cita` script, that would be more perfect:
> 
> * Generating chain configuration, just likes:
> 
> ```
> ./env.sh ./scripts/create_cita_config.py create --super_admin "0x4b5ae4567ad5d9fb92bc9afd6a657e6fa13a2523"  --chain_name test2-chain --jsonrpc_port 2337 --ws_port 5337 --grpc_port 6000 --nodes "127.0.0.1:8000,127.0.0.1:8001,127.0.0.1:8002,127.0.0.1:8003"
> ```
> * Exposing a port from the running docker which likes `-p 1337:1337`.
> 
> After that, we can get a `cita-block-chain` by using `cita` which just installed.

## How about these?

__For the first feature:__

`cita create` has these arguments:

|Arguments|Description|
|-|-|
|authorities|Authorities (addresses) list|
|chain_name|Name of the new chain|
|nodes|Node network addresses for new nodes|
|super_admin|Address of super admin||
|contract_arguments|Update constructor arguments for system contract.|
|time_stamp|Specify a timestamp to use|
|grpc_port|grrpc port for this chain|
|jsonrpc_port|jsonrpc port for this chain|
|ws_port|websocket port for this chain|
|enable_tls|The data is encrypted and transmitted on the network|

_Arguments might too long that not friendly for the command-line._

I have an idea, once we run `cita`, we create a `cita-block-chain` with preset arguments by default, and, the `create` command will interact with users line by line: 
```shell
$ cita create "(super_admin)"
# is super_admin the only required argument?
$ cita create -v "(super_admin)"
# the verbose tag `-v`.
please input your configure arguments, press `enter` will set default:
$ chain name: 
# after input, a new line comes or replaces the upper line's view...
$ json_rpc_port: 
...
```

I think it is no good to rewrite this method in `shell` from `python` now? Should I call the `create_cita_chain.py` from `./cita` directly? the better might be write a rust commander directly? XD

__The second feature:__

|Command|Execution|
|-|-|
|`cita` | run docker container as `-p 1337:1337`|
|`cita port 6666:6666`|run docker container as `-p 6666:6666`|

leeyr338 replied:
- For the first feature, I think is ok to call `create_cita_chain.py` from `./cita` directly.
- And the second feature, your proposal looks good!

kaikai1024 replied:
> the better way might be writing a rust commander directly

yeah, That's the plain. #221 We can discuss this on the issue.

>  Should I call the create_cita_chain.py from ./cita directly

About this, I'd like to seperate the create_cita_config as an independent tool. The reason is that the tool should be just used to create genesis, the genesis is actually what cita needs.
I think cita does not need to consider how genesis is created. so we can have py version, Rust version, something like that.

leeyr338 replied:
Independent tool sounds good!

So we also need  `brew install cita-config`(or some command like that)  to help newcomer running CITA easily.

clearloop replied:
I've written the `port` command just now, 

A new bug occurs, and I wondering why:
```
2019-03-21T11:35:25.543589700+08:00 network connect b93f89efcac12d3f8cb2002e55b1b7df26173aaf968115fd7e376b1021187506 (container=778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca, name=bridge, type=bridge)
2019-03-21T11:35:26.017352200+08:00 container start 778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca (image=cita/cita-run:ubuntu-18.04-20181009, maintainer=Cryptape Technologies <contact@cryptape.com>, name=cita)
2019-03-21T11:35:26.197330600+08:00 container die 778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca (exitCode=0, image=cita/cita-run:ubuntu-18.04-20181009, maintainer=Cryptape Technologies <contact@cryptape.com>, name=cita)
2019-03-21T11:35:26.475652000+08:00 network disconnect b93f89efcac12d3f8cb2002e55b1b7df26173aaf968115fd7e376b1021187506 (container=778b4bf6230f3bd48edc0ddafa051f579c9e58e5b3e932aaa1946ca751ed27ca, name=bridge, type=bridge)
```

Here is my config code:
```
# Run container                                                                                                                                                                                             
docker ps | grep ${CONTAINER_NAME} > /dev/null 2>&1
if [ $? -ne 0 ]; then
    echo "Start docker container ${CONTAINER_NAME} ..."
    docker rm ${CONTAINER_NAME} > /dev/null 2>&1
    docker run -d \
           --net=${SYSTEM_NET} \
           --volume ${SOURCE_DIR}:${WORKDIR} \
           --volume ${LOCALTIME_PATH}:/etc/localtime \
           --volume ${HOME}/.docker_cargo/git:${CARGO_HOME}/git \
           --volume ${HOME}/.docker_cargo/registry:${CARGO_HOME}/registry \
           --env USER_ID=${USER_ID} \
           --workdir ${WORKDIR} \
           --name ${CONTAINER_NAME} \
           -p $EXPOSE ${DOCKER_IMAGE} \
           /bin/bash -c "${INIE_CMD}"
    # Wait entrypoint.sh to finish                                                                                                                                                                          
    sleep 3
fi
```

__About the `brew install cita`:__
Should just install cita compiled code without docker?

Docker environment is more like production, for company, developers just debug their code and learning cita on it.

Docker image is more than 2GB, but __compiled code zips__  are less than 50MB, it's a favor for newcomers who just want to try it.

__Sperate `cita-config`:__

It's really nice, I'll try it after fixed the bug above, 

btw, I'll also inject the create code into `cita` as just like `create_cita_config.py` does, the line-by-line interactive, if we got `cita-config`, it becomes verbose.

leeyr338 replied:
Without docker, means that we need to release compiled code for different OS. There are a lot of work for test!

kaikai1024 replied:
Could we discuss it on your issue? so someone else can join it, too.

clearloop replied:
> Could we discuss it on your issue? so someone else can join it, too.

Of course, the `Makefile` has already outdated, it wouldn't be sad...XD

> Without docker, means that we need to release compiled code for different OS. There are a lot of work for test!

It is a problem, but, the developing experience is really important, for example:

__Why ethereum got such many developers and users?__

Except the business operating and the smart-contract concepts, even publishing coin functions, one of the reasons on the core thing, the programs, they do have great programs, easy to use, easy to downloads.

Developers are happy to use it,  the developing process endues them senses of accomplishment, the hope of the future, they talking about it, writing articles about it, publishing programs about it —— more and more people join.

Every CITA show in the public view from company usage,  it is reach to the Ethereum, there are big needs and markets in China, the Chinese people, who expecting the changing of the time, who enjoy new technologies, who are dreaming about the future the whole lives, somebody has faith in blockchain, they do need to believe, something is happening.

emmmm, anyway, I think it is important, but I am not sure if it is the most, maybe the `different OS` testing problem can be delayed to `cita-1.0.0-stable`? 

It is quite nice to add: 

+ brew
+ snap

or

+ Linux-basic-version
+ MacOS-basic-version
+ Windows-basic-version

in the current downloads list.

kaikai1024 commented:
Are you still working on this? Is there something i can help?

clearloop commented:
> Are you still working on this? Is there something i can help? @clearloop

I'd like to write a formula or snap to make cita installing and using easier, I've made the formula works in my branch, but they should depend on official resources

In current commit, I modified
+ `cita/env.sh`
+ `cita/scripts/cita`
+ `cita/scripts/release.sh`

and deleted
+ `cita/daemon.sh`
+ `cita/scripts/env.sh`
+ `cita/scripts/daemon.sh`

Users can use `cita`, `cita-env` and  `cita-config` in global environment in my scripts.

And in current version, `cita services` and `cita-cli` are separated, it's quite clean, `cita-services` no need to nest too many commands just like `geth`.

I don't know how to continue, I can't change too many files once time, if I did it, the scripts will deviate from the main track.  

So I wrote a dapp on cita last days XD

kaikai1024 commented:
> I'd like to write a formula or snap to make cita installing and using easier, I've made the formula works in my branch, but they should depend on official resources

I will check this later.

Now, let's focus on this pr and fix the CI.

kaikai1024 commented:
> I don't know how to continue, I can't change too many files once time, if I did it, the scripts will deviate from the main track.

The CI is required for merging.  Or is there some way to sperate it to many little ones? Can you read the info about the circle and travis? If not, you can try to run ci on your local pc. or i can help to fix it.

kaikai1024 commented:
> Users can use cita, cita-env and cita-config in global environment in my scripts.

That's what i want to do, too.

clearloop commented:
> The CI is required for merging. Or is there some way to sperate it to many little ones? Can you read the info about the circle and travis? If not, you can try to run ci on your local pc. or i can help to fix it.

Oops, it was my fault, I forgot to modify the CI's code to build tests, I will optimize my code again and make it through myself, soon, thx for your reminder~

kaikai1024 commented:
cool, I will help to fix the travisCI.

clearloop commented on scripts/cita:
```diff
@@ -2,17 +2,32 @@
+CITA_SCRIPTS="$(dirname $CITA_BIN)/scripts"
+if [ "$1" != "bebop" ]; then
+    if [ -e $CITA_BIN/cita-env ];then
+        $CITA_BIN/cita-env bin/cita bebop $@
```
The `bebop` keyword is my selfish addition, it's from a famous cartoon in 1998, [cowboy bebop][1], galaxy adventure, Rock and Jazz, sounds like [this][2].

If I need to modify it to another expression, it's okay.

[1]: https://en.wikipedia.org/wiki/Cowboy_Bebop
[2]: https://music.163.com/#/song?id=592687

kaikai1024 replied:
so `bebop` is used for using docker, right?

I'd like to use your choice on this pr for thanks of your contribution.

Btw, Maybe will change it to a more easy-understanding word in the future, i'm not suer about this.

clearloop replied:
> so bebop is used for using docker, right?
Yep, `bebop` is only for commands inside docker and users using cita through docker may never reach it.

> Btw, Maybe will change it to a more easy-understanding word in the future, i'm not suer about this.
It's fine, thank you.

clearloop commented on env.sh:
```diff
@@ -1,71 +1,79 @@
-    else
-        INIT_CMD="echo -e '[source.crates-io]\nregistry = \"https://github.com/rust-lang/crates.io-index\"\nreplace-with = \"ustc\"\n[source.ustc]\nregistry = \"https://mirrors.ustc.edu.cn/crates.io-index\"' | sudo tee /opt/.cargo/config;while true;do sleep 100;done"
-    fi
-
```
I not sure about these lines, I built failed when I use `ustc` sources, the same issue can refer to [#6513][1] in cargo.

[1]: https://github.com/rust-lang/cargo/issues/6513

kaikai1024 replied:
Would you please to create another issue about this.
I think it's better to fix it on another pr.

kaikai1024 commented on .travis.yml:
```diff
@@ -54,11 +54,11 @@ jobs:
         - rm -rf /opt/cita-run/test-chain
         - cd $TRAVIS_BUILD_DIR/target/install
-        - ./env.sh ./scripts/create_cita_config.py create
+        - ./bin/cita create
```
maybe use bebop?

clearloop replied:
Might not, 

The official branch use `env.sh`, 

And, every command needs `./env.sh` or `daemon.sh` converts to use `bebop` or `bebop daemon` automatically in my branch XD, 

`bebop` keyword only required when scripts running directly without docker.

clearloop replied:
oops, the No.55 line, seems the commands are in docker?

kaikai1024 replied:
> bebop keyword only required when scripts running directly without docker.

this is the situation

kaikai1024 replied:
I notice that the failed job is still using the old config.

Would you please to rebase the latest branch and squash some commits.

clearloop replied:
I'm afraid I couldn't do it well, I'm a bad __rebaser__, last time rebasing exchanging my 3 force push....XD, but I will try it again anyway.

kaikai1024 replied:
You are right, No need to use `bebop`

kaikai1024 replied:
> I'm afraid I couldn't do it well, I'm a bad **rebaser**, last time rebasing exchanged my 3 force push....XD, but I will try it again anyway.

remeber to use `rebase -i develop`. you can try to keep your commit clean or maybe we can use the squash merging.

clearloop replied:
Hey @kaikai1024 , could you checkout my [testing][1] branch, if I did the right thing, should I do the same to my develop branch, and __force push_...again?

[1]: https://github.com/clearloop/cita/commits/testing

clearloop replied:
If the concrete commands I should use are:

```
git rebase -i Head~9
# ~> squash the commits
# ~> do some commit describes
# ~> and then, my branch became clean.
git push -f
```

kaikai1024 replied:
I'm not sure about the remote info
1. check it using  `git remote -v`(assume that cryptape/cita is origin)
2. `git fetch origin/develop`
3. `git rebase -i origin/develop`

kaikai1024 replied:
Could i clone your repo and creae a new branch on my repo, you can check the commits.

kaikai1024 replied:
We can do it later. let's focus on the travisCI.

clearloop replied:
Okay, but it seems I just solve the squash problem, [here][1] it is.

All I confused is if I have to __force push__.

The ci's problem is all about the JSONRPC, no response, I've written the condition about `--net=host` when it's not running on macos, and I'm thinking about why it happens now.

[1]: https://github.com/clearloop/cita/commits/testing

kaikai1024 replied:
ok.

I will try to fix the travis.
Actually, you just run the travis on your own repo, it's much faster

kaikai1024 replied:
ok.
I am trying to fix it on my branch:
https://github.com/kaikai1024/cita/tree/fix_travis_cita

* rebase the later develop
* fixup other commits
* reword the commit

kaikai1024 replied:
> Hey @kaikai1024 , could you checkout my testing branch?

That's right

clearloop replied:
CI's scripts copies the `target/install-cache` into `target/install` after deleting it:
```
 before_install:
    - cd $TRAVIS_BUILD_DIR
    - rm -rf target/install
    - cp -rv target/release-cache target/install
```

The error might about the exec grant of the copied `cita-env`, 

Here are the logs of my latest build: 
```
$ ./bin/cita start test-chain/0
cita-env exists
error: exec: "/home/travis/build/clearloop/cita/target/install/bin/cita-env": stat /home/travis/build/clearloop/cita/target/install/bin/cita-env: no such file or directory
```

I wrote exists' tests, The file is exist, but it isn't executable, so I find this same [question][1].

I'm quite confused now, `cita` has the executable grant, why `cita-env` and `cita-config` not? even I `chmod +x` them in `release.sh`.

[1]: https://askubuntu.com/questions/133389/no-such-file-or-directory-but-the-file-exists

kaikai1024 replied:
Maybe something with the cache of travis.
Wait a minute, I'll try it.

kaikai1024 replied:
I have the same problem on my pc.

kaikai1024 commented on env.sh:
```diff
@@ -1,71 +1,79 @@
+SOURCE_DIR="$(cd $(dirname "$0") && pwd -P)"
+
+test -f "${SOURCE_DIR}/CODE_OF_CONDUCT.md"
+if [ $? -eq 0 ]; then
```
nice

kaikai1024 commented on scripts/release.sh:
```diff
@@ -55,9 +56,15 @@ cp -f   scripts/create_cita_config.py      target/install/scripts/
 cp -f scripts/amend_system_contracts.sh    target/install/scripts/
 cp -f scripts/amend_system_contracts.py    target/install/scripts/
+
+# 8) delete building container
```
Why delete the container at this script?

clearloop replied:
The build container is just for build, developers will not need it when building is completed.

kaikai1024 replied:
ok. i misunderstand it

kaikai1024 commented on env.sh:
```diff
@@ -1,71 +1,79 @@
-if [ $# -gt 0 ]; then
+# Start nodes outside container directly
+# if `daemon.sh` just use in start node?
+# Remote Commands:
```
Is `bin/cita start test-chain/0` right according to the comments?

clearloop replied:
`start` is a special command, every `start` parameter inputs, cita will run daemon automatically. 

here is my latest log:
```
$ ./bin/cita start test-chain/0
cita receive parameters: start test-chain/0
env.sh recive paraemters: /home/travis/build/clearloop/cita/target/install/bin/cita-env bin/cita bebop --daemon start test-chain/0
error: exec: "/home/travis/build/clearloop/cita/target/install/bin/cita-env": stat /home/travis/build/clearloop/cita/target/install/bin/cita-env: no such file or directory
```
The `cita-env` did run, but the error still coms...XD

kaikai1024 replied:
```
+++ dirname bin/cita
++ cd bin
++ pwd -P
+ CITA_BIN=/home/kaikai/Work/cita/target/install/bin
++ dirname /home/kaikai/Work/cita/target/install/bin
+ CITA_SCRIPTS=/home/kaikai/Work/cita/target/install/scripts
+ '[' start '!=' bebop ']'
+ '[' -e /home/kaikai/Work/cita/target/install/bin/cita-env ']'
+ /home/kaikai/Work/cita/target/install/bin/cita-env bin/cita bebop start test-chain/0
error: exec: "/home/kaikai/Work/cita/target/install/bin/cita-env": stat /home/kaikai/Work/cita/target/install/bin/cita-env: no such file or directory
```

clearloop replied:
And, the `cita start` command in my branch, can run well while building locally or replaced them into the official compiled code, invoked into a formula.

clearloop replied:
I think I might know what happened...I optimized the code in compiling scripts, but not update it in my branch.

clearloop replied:
I raised the bug, after fixing `circles` —— messed up the `SOURCE_DIR`.

kaikai1024 commented:
TODO: the doc(you can update the doc at the same time)

***squash merging*** 

@yangby-cryptape *need yang's approve, please take a little time to take a look*

it's better to sort out your commits.

clearloop commented:
> TODO: the doc(you can update the doc at the same time)
> 
> _**squash merging**_
> 
> @yangby-cryptape _need yang's approve, please take a little time to take a look_
> 
> it's better to sort out your commits.

Should I squash and force push now?

kaikai1024 commented:
> Should I squash and force push now?

It's up to you.

My job is to keep the commits of develop clean. That's means i will use squash merging(package the commits to a commit) when i think the commits is a little messy or the pr is should be one commit.

nice job. just wait for another approval!

kaikai1024 commented:
And, would you like to contribute to the doc about the part that you changed?

please let me know and then i decide to create an issue or not.

thank you. :heart:

kaikai1024 commented:
Have you rebased the latest develop? and updated the submodule? 

the bft and forever should not be updated.

clearloop commented:
> Have you rebased the latest develop? and updated the submodule?
> 
> the bft and forever should not be updated.

yep，I'll write the doc and try to fix the rust code if I can, anyway, I'm a rust developer...

kaikai1024 commented:
> I'm a rust developer.

:crab:

kaikai1024 commented:
I will update the doc. @clearloop https://github.com/cryptape/citahub-docs/issues/115
